Refactor Schema::Printer to use Language::Printer

xuorig opened the pull request:
This PR does a few things:

  - Adds a lot of options to `DocumentFromSchemaDefinition`:
    - context: nil
    - only: nil
     - except: nil
     - include_introspection_types: false
     - include_built_in_directives: false
     - include_built_in_scalars: false, 
     - always_include_schema: false

This allows printers to customize the AST before printing it, and removes this logic from printers which become "dumb".

  - Refactors `SchemaPrinter` to use `DocumentFromSchemaDefinition`, this 🔥 a lot of code.
    - SchemaPrinter now inherits from `Language::Printer`.
    - Makes it much easier to extend now

   - Fixes a few issues with `Language::Printer` for printing schemas.


I tried to keep the Schema::Printer API exactly the same, only the underlying logic is changing.

xuorig commented:
cc @rmosolgo this is ready for review!

(not sure what is going on with codeclimate)

rmosolgo commented:
Cool! I'll take a close look soon.

Code Climate was _kinda_ broken before, now it looks _very_ broken 😬

rmosolgo commented on lib/graphql/language/document_from_schema_definition.rb:
```diff
@@ -18,18 +36,18 @@ def document
         end
 
-        if schema.subscription
+        if schema.subscription && warden.get_type(schema.subscription.name)
```
How about using [`#root_type_for_operation("subscription")`](https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/schema/warden.rb#L118) here and for `mutation` and `query` ?

xuorig replied:
What's the difference here? It doesnt look like `root_type_for_operation` uses the warden 🤔

rmosolgo replied:
Oh, to specify, `Warden#root_type_for_operation`, not `Schema#root_type_for_operation`: 

https://github.com/rmosolgo/graphql-ruby/blob/20d70eb374809835fedaf85bd8bffd0d71799abe/lib/graphql/schema/warden.rb#L118-L125

xuorig replied:
ohhhhhh 💡 awesome.

Done :)

rmosolgo commented on lib/graphql/language/document_from_schema_definition.rb:
```diff
@@ -6,10 +6,28 @@ module Language
+    # @param except [<#call(member, ctx)>]
+    # @param include_introspection_types [Boolean] Wether or not to include introspection types in the AST
+    # @param include_built_in_scalars [Boolean] Wether or not to include built in scalars in the AST
+    # @param include_built_in_directives [Boolean] Wether or not to include built in diirectives in the AST
```
W**h**ether 😊

rmosolgo commented on lib/graphql/schema.rb:
```diff
@@ -564,6 +564,12 @@ def to_json(*args)
       JSON.pretty_generate(as_json(*args))
     end
 
+    def respects_root_name_conventions?
```
Is this concern only for printing the schema? If so, can we keep it out of `schema.rb`? There's already a lot going on in here, and if we can get away with putting it somewhere else, I'd like to do that!

xuorig replied:
done!

rmosolgo commented on spec/graphql/schema/printer_spec.rb:
```diff
@@ -390,7 +389,7 @@
 
 enum Choice {
   BAR
-  BAZ @deprecated(reason: "Use \\\"BAR\\\".")
```
🤔  Can you help me understand why these escapes are not needed anymore? (Were they wrong to begin with?)

xuorig replied:
Apparently it has to do with the call to inspect here: https://github.com/rmosolgo/graphql-ruby/blob/master/lib/graphql/schema/printer.rb#L164

```
irb(main):031:0> field_or_enum_value.deprecation_reason
=> "use \"this\""
irb(main):032:0> field_or_enum_value.deprecation_reason.to_s
=> "use \"this\""
irb(main):033:0> field_or_enum_value.deprecation_reason.to_s.inspect
=> "\"use \\\"this\\\"\""
```

I don't think its needed 🤔

rmosolgo replied:
Cool, thanks for the details!

xuorig commented:
Addressed your comments, few other things I worry about:

  - Were these printer modules used by users? 😨 (might be an acceptable breaking change..)

  - The schema printer now only has access to AST nodes and not schema members definitions. This means printing according to something in metadata for example is now much more complex. I suspect there's not that many valid use cases here. Information would usually find its way in AST, especially as we start maybe accepting custom directives. Worst case we can open up `DocumentFromSchemaDefinition` so users can have custom ASTs from schema definitions.

In the end this last point might be a good thing since you can only print valid schema ASTs, making your schema readable by any graphql implementation.

rmosolgo commented on lib/graphql/language/document_from_schema_definition.rb:
```diff
@@ -18,73 +36,81 @@ def document
+          field_node.directives << GraphQL::Language::Nodes::Directive.new(
+            name: GraphQL::Directive::DeprecatedDirective.name,
+            arguments: [GraphQL::Language::Nodes::Argument.new(name: "reason", value: field.deprecation_reason)]
+          )
```
👌

rmosolgo commented on lib/graphql/schema/printer.rb:
```diff
@@ -74,277 +85,37 @@ def self.print_schema(schema, **args)
+      end
 
-      private_constant :IS_USER_DEFINED_MEMBER
+      def print_deprecated(directive)
```
Can this method be private? Or, better yet, can this code be inlined into `print_directive`? (I couldn't see any usages of it, but maybe I overlooked them!)

xuorig replied:
good point, fixed!

rmosolgo commented on spec/graphql/schema/build_from_definition_spec.rb:
```diff
@@ -520,7 +520,7 @@ def build_schema_and_compare_output(definition)
 
 type Mutation {
-  hello(str: String, int: Int, color: Color = RED, nullDefault: Int = null): String
+  hello(color: Color = RED, int: Int, nullDefault: Int = null, str: String): String
```
yay, sorting

rmosolgo commented:
🙌  🎉

xuorig commented:
🎉
